### PR TITLE
feat(expressive): add expressive variation for the accordion and button

### PR DESCRIPTION
**Pull request by annawen1**
Closes: 
[Create RFC for an expressive moment in a Carbon core component #1782](https://github.com/carbon-design-system/ibm-dotcom-library/issues/1782)

#### Description
This is an RFC for an approach the IBM.com team has suggested to introduce the expressive theming. This approach creates an expressive variant of a carbon component that would users to set the expressive theming per component.

Previously the IBM.com team created a proxy that would allow for expressive theming to be applied to carbon's core components when a flag is turned on. A major issue with this approach was the maintainability - every time the carbon dependencies were updated in the IBM.com repo we would have to comb through all components and ensure the expressive styles were still intact. 

With this RFC approach, it would be easier to maintain the expressive styles as the styles are built into the core component itself. 

#### Changelog

**New**

- `expressive` prop to be passed to accordion and set a `--accordion--expressive` modifier classname
- add expressive styles under the new modifier class
- `expressive` knob for storybook accordion story

#### Testing / Reviewing

Apr-02-2020 10-50-51

Steps to Review:
1. go to Accordion story on storybook
2. check the `expressive theme` knob
3. Notice the expressive styling changes that appear

### CODE REVIEW

**Comment by annawen1** [comment]
@oliviaflory @jeffchew @wonilsuhibm @andysherman2121

**Comment by aagonzales** [comment]
Oh nice! Yes, this is what I would expect if we go the variant route. I don't think I would call the knob "expressive theme" though maybe just "expressive" for a variant. Is it actually inline calling another theme or is it just changing the values? It looked like it was just changing the type token to be `body-long-02` which wouldn't be a theme switch?

**Comment by oliviaflory** [comment]
> I don't think I would call the knob "expressive theme" though maybe just "expressive" for a variant. Is it actually inline calling another theme or is it just changing the values? It looked like it was just changing the type token to be `body-long-02` which wouldn't be a theme switch?

Agreed we could go with "expressive" for the knob. It is changing the type token to be `body-long-02` and altering the container height and icon size. So not really a new theme, but a variant of the component.


These are the other components we'd imagine having an expressive variants:
- accordion
- button
- link
- list
- structured list
- tabs
- tile

File: packages/components/src/components/accordion/_accordion.scss [hunk 1]
```diff
@@ -197,6 +197,25 @@
   .#{$prefix}--accordion__title.#{$prefix}--skeleton__text {
     margin-bottom: 0;
   }
+
+  .#{$prefix}--accordion--expressive {
+    overflow: hidden;
+
+    .#{$prefix}--accordion__heading {
```
Comment (tw15egan) [inline packages/components/src/components/accordion/_accordion.scss]: Can we unnest these selectors? We're trying to reduce nesting throughout the codebase so it's easier to search for specific selectors in the future, e.g

```scss
.#{$prefix}--accordion--expressive .#{$prefix}--accordion__heading {
    padding: $carbon--spacing-04 0;
}
```

File: packages/components/src/components/accordion/_accordion.scss [hunk 2]
```diff
@@ -197,6 +197,25 @@
...
+      padding: $carbon--spacing-04 0;
+    }
+
+    .#{$prefix}--accordion__title,
+    .#{$prefix}--accordion__content > p {
+      @include type-style('body-long-02');
+    }
+
+    .#{$prefix}--accordion__arrow {
```
Comment (tw15egan) [inline packages/components/src/components/accordion/_accordion.scss]: See above ☝️

File: packages/components/src/components/accordion/_accordion.scss [hunk 3]
```diff
@@ -197,6 +197,25 @@
...
+
+  .#{$prefix}--accordion--expressive {
+    overflow: hidden;
+
+    .#{$prefix}--accordion__heading {
+      padding: $carbon--spacing-04 0;
+    }
+
+    .#{$prefix}--accordion__title,
+    .#{$prefix}--accordion__content > p {
```
Comment (tw15egan) [inline packages/components/src/components/accordion/_accordion.scss]: See above ☝️

File: packages/components/src/components/accordion/_accordion.scss (hunk 1 above)
Comment (annawen1) [reply packages/components/src/components/accordion/_accordion.scss]: sure, will unnest these selectors

File: packages/react/src/components/Accordion/Accordion-story.js [hunk 4]
```diff
@@ -39,7 +39,8 @@ storiesOf('Accordion', module)
           'Accordion heading alignment (align)',
           ['start', 'end'],
           'end'
-        )}>
+        )}
+        expressive={boolean('expressive', false)}>
```
Comment (emyarod) [inline packages/react/src/components/Accordion/Accordion-story.js]:
```suggestion
        expressive={boolean('Expressive variation (expressive)', false)}>
```

to match the knob labels across the storybook

**Review by emyarod — COMMENTED** [review]
I'm a bit out of the loop on this but are expressive component variations going to be a Carbon level setting?

**Comment by annawen1** [comment]
> I'm a bit out of the loop on this but are expressive component variations going to be a Carbon level setting?

I don't believe so...it'll be applied on a per component basis... @jeffchew is that right?

**Comment by jeffchew** [comment]
> > I'm a bit out of the loop on this but are expressive component variations going to be a Carbon level setting?
> 
> I don't believe so...it'll be applied on a per component basis... @jeffchew is that right?

Yes, we spoke about having expressive "moments" in the component level.

**Comment by emyarod** [comment]
oh I see, by Carbon level setting I meant whether or not it would live in this repository, but it seems the answer to that is also yes

**Review by mattrosno — COMMENTED** [review]
**1) Developer experience** - if every instance of a Carbon component used on dotcom needs an expressive prop set to true, how do you enforce that? Or, how do you make it easy for your developers to not forget setting an expressive prop to true on every component they use?

**2) Documentation** - for the website component documentation, do we need to set URL parameters based on the current props you have configured, so you can direct-link to a component with expressive turned on? Also, should we persist certain "global" props like expressive so if you turn on expressive for accordion, then view button, it's expressive is also on by default?

**3) Naming** - is `expressive` as a boolean right here? Or maybe something like `typeTheme="expressive"` to make this more scalable should there be something beyond default (productive) and expressive in the future? Or maybe `<Button moment="expressive">` if moment is a better word?

**4) Sketch kit** - if provided by Carbon core, this needs to be available in our Sketch kits. Take button for example.

image

4 themes x 5 kinds x 3 sizes x 3 icon x 5 states = 900 button symbols today. If there's a new expressive prop, we're looking at 900 x 2 = 1,800 symbols that we need to maintain, just for button!

My point - a new prop for this doubles the number of permutations that we need to test for and provide in our Sketch kit. Even though adding a new prop, modifier class, and modifier declaration blocks are pretty minimal... significantly increasing our test coverage and Sketch kit symbols is concerning both in initial effort and ongoing maintenance for everything we do going forward.

**5) Completeness** - I'm all for incremental design and development, but if we start this, we need to be committed to completing it. I don't want us in a scenario like our UI Shell that takes a ~year to be color-themable like every other component that we support.

**Comment by joshblack** [comment]
Hey there! Thanks so much for putting all of this work together 🙏  Going to go ahead and close this one out since it seems like the conversation might have moved to another PR, but let me know if that's not the case and we can re-open!

**Pull request closed without merging**